### ISSUE TITLE
Populated documents are flattened to ObjectIds when assigned to another document

### ISSUE THREAD

**Issue by neubee** [Automattic/mongoose#14418]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.12.6

### Node.js version

16.20.2

### MongoDB server version

5.0.25

### Description

I'm having a problem where Mongoose will "flatten" documents to just an ObjectId when assigned to other documents. It's been difficult to narrow down when exactly this behavior occurs, but I do know that it will happen when populating an array and then assigning one, or all, of those populated documents to an array in another document. There could be other cases were this happens as well that I haven't found.

I'm not certain if this is a bug or expected behavior, but this is causing a lot of trouble in my project. Also, while I've always had trouble with this type of Mongoose behavior it started happening in new areas of my code when I updated from Mongoose 5.13.20 to 6.12.6

If this is expected behavior, how can I stop it from happening? Right now I need to resort to keeping/passing around the populated version of the documents or I need to populate them again after assigning them to the second parent.

### Steps to Reproduce

```js
const mongoose = require("mongoose");
const { MongoMemoryServer } = require("mongodb-memory-server");

(async () => {
  const mongoServer = await MongoMemoryServer.create();
  await mongoose.connect(mongoServer.getUri(), { dbName: "test" });

  const cartSchema = new mongoose.Schema({
    products: [
      {
        product: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Product",
        },
        quantity: Number,
      },
    ],
    singleProduct: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
    },
  });
  const purchaseSchema = new mongoose.Schema({
    products: [
      {
        product: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Product",
        },
        quantity: Number,
      },
    ],
    singleProduct: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
    },
  });

  const productSchema = new mongoose.Schema({
    name: String,
  });

  const Cart = mongoose.model("Cart", cartSchema);
  const Purchase = mongoose.model("Purchase", purchaseSchema);
  const Product = mongoose.model("Product", productSchema);

  const dbProduct = new Product({
    name: "Bug",
  });
  await dbProduct.save();

  const dbCart = new Cart({
    products: [
      {
        product: dbProduct,
        quantity: 2,
      },
    ],
    singleProduct: dbProduct,
  });
  await dbCart.save();

  const foundCart = await Cart.findById(dbCart._id).populate(
    "products.product singleProduct",
  );
  console.log("foundCart:");
  console.log(foundCart);

  // Seems like it only is an issue with arrays, singleProduct is not flattened
  const purchaseFromDbCart = new Purchase({
    products: foundCart.products,
    singleProduct: foundCart.singleProduct,
  });
  console.log("purchaseFromDbCart:");
  console.log(purchaseFromDbCart);
  /*
  {
    products: [
      {
        product: new ObjectId("65e8dc15e063a7ea53303f62"),
        quantity: 2,
        _id: new ObjectId("65e8dc16e063a7ea53303f65")
      }
    ],
    singleProduct: {
      _id: new ObjectId("65e8dc15e063a7ea53303f62"),
      name: 'Bug',
      __v: 0
    },
    _id: new ObjectId("65e8dc16e063a7ea53303f6b")
  }
  */

  await mongoose.disconnect();

  // When DB is not involved the full document is kept
  const nonDbProduct = new Product({
    name: "Bear",
  });
  const nonDbCart = new Cart({
    products: [
      {
        product: nonDbProduct,
        quantity: 2,
      },
    ],
    singleProduct: nonDbProduct,
  });
  const nonDbPurchase = {
    products: nonDbCart.products,
    singleProduct: nonDbCart.singleProduct,
  };
  console.log("nonDbPurchase:");
  console.log(nonDbPurchase);

  process.exit();
})();
```

**Comment by neubee** [comment]
So weird! Just to be sure, what do you see when using this slightly modified code (less stuff put to console)
```
const mongoose = require("mongoose");
const { MongoMemoryServer } = require("mongodb-memory-server");

(async () => {
  const mongoServer = await MongoMemoryServer.create();
  await mongoose.connect(mongoServer.getUri(), { dbName: "test" });

  const cartSchema = new mongoose.Schema({
    products: [
      {
        product: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Product",
        },
        quantity: Number,
      },
    ],
    singleProduct: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
    },
  });
  const purchaseSchema = new mongoose.Schema({
    products: [
      {
        product: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Product",
        },
        quantity: Number,
      },
    ],
    singleProduct: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
    },
  });

  const productSchema = new mongoose.Schema({
    name: String,
  });

  const Cart = mongoose.model("Cart", cartSchema);
  const Purchase = mongoose.model("Purchase", purchaseSchema);
  const Product = mongoose.model("Product", productSchema);

  const dbProduct = new Product({
    name: "Bug",
  });
  await dbProduct.save();

  const dbCart = new Cart({
    products: [
      {
        product: dbProduct,
        quantity: 2,
      },
    ],
    singleProduct: dbProduct,
  });
  await dbCart.save();

  const foundCart = await Cart.findById(dbCart._id).populate(
    "products.product singleProduct",
  );

  const purchaseFromDbCart = new Purchase({
    products: foundCart.products,
    singleProduct: foundCart.singleProduct,
  });
  console.log('Product in cart array:')
  console.log(foundCart.products[0].product.name) // I see the string "Bug"
  console.log('Product in purchase array:')
  console.log(purchaseFromDbCart.products[0].product.name) // I see "undefined"
```

**Comment by neubee** [comment]
Did a bit more testing here and I have two additional notes, in case they are helpful.

1. I'm seeing the same issue in version 8.2.1 of Mongoose
2. It looks like the same behavior happens when assigning an object from a populated array to a non-array property. Building off the example I already provided, if you assign `foundCart.products[0].product` to a `purchase.singleProduct` it will also be flattened into an ObjectId
```
  const purchaseWithSingleProductAssignedFromProductsArray = new Purchase({
    singleProduct: foundCart.products[0].product,
  })
  console.log(purchaseWithSingleProductAssignedFromProductsArray.singleProduct) // I see "new ObjectId("some object ID here")"
```

**Comment by IslandRhythms** [comment]
Confirmed bug with that reproduction script, and is present on 8.2.1. of mongoose

**Comment by vkarpov15** [comment]
Investigating this issue, looks like it may be caused by the fix for #11442.
